commit c81e542b32815effe9d107da378e6010e8a0b415
Author: John Spray <jspray@redhat.com>
Date:   Mon Jul 4 15:22:27 2016 +0100

    mgr: enable multiple python modules
    
    serve() each one in a separate thread, include a shutdown()
    hook so that we can tear down cleanly.
    
    Signed-off-by: John Spray <john.spray@redhat.com>

diff --git a/src/mgr/MgrPyModule.cc b/src/mgr/MgrPyModule.cc
index c1dee9f1968..4e8d0d3bf64 100644
--- a/src/mgr/MgrPyModule.cc
+++ b/src/mgr/MgrPyModule.cc
@@ -93,6 +93,27 @@ int MgrPyModule::serve()
   return 0;
 }
 
+// FIXME: DRY wrt serve
+void MgrPyModule::shutdown()
+{
+  assert(pClassInstance != nullptr);
+
+  PyGILState_STATE gstate;
+  gstate = PyGILState_Ensure();
+
+  auto pValue = PyObject_CallMethod(pClassInstance,
+      (const char*)"shutdown", (const char*)"()");
+
+  if (pValue != NULL) {
+    Py_DECREF(pValue);
+  } else {
+    PyErr_Print();
+    derr << "Failed to invoke shutdown() on " << module_name << dendl;
+  }
+
+  PyGILState_Release(gstate);
+}
+
 void MgrPyModule::notify(const std::string &notify_type, const std::string &notify_id)
 {
   assert(pClassInstance != nullptr);
diff --git a/src/mgr/MgrPyModule.h b/src/mgr/MgrPyModule.h
index 8a5dc94edef..dd490d7e740 100644
--- a/src/mgr/MgrPyModule.h
+++ b/src/mgr/MgrPyModule.h
@@ -46,7 +46,6 @@ private:
   PyObject *pClass;
   PyObject *pClassInstance;
 
-
   std::vector<ModuleCommand> commands;
 
   int load_commands();
@@ -57,6 +56,7 @@ public:
 
   int load();
   int serve();
+  void shutdown();
   void notify(const std::string &notify_type, const std::string &notify_id);
 
   const std::vector<ModuleCommand> &get_commands() const
@@ -64,6 +64,11 @@ public:
     return commands;
   }
 
+  const std::string &get_name() const
+  {
+    return module_name;
+  }
+
   int handle_command(
     const cmdmap_t &cmdmap,
     std::stringstream *ss,
